xrootd/xrdfuse/xrdfuse.go

```diff
@@ -0,0 +1,166 @@
+// Copyright 2018 The go-hep Authors.  All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// Package xrdfuse contains implementation of a FUSE api for the XRootD.
+package xrdfuse
+
+import (
+	"context"
+	"log"
+	"os"
+	"path"
+
+	"github.com/hanwen/go-fuse/fuse"
+	"github.com/hanwen/go-fuse/fuse/nodefs"
+	"github.com/hanwen/go-fuse/fuse/pathfs"
+	"go-hep.org/x/hep/xrootd/client"
+	"go-hep.org/x/hep/xrootd/xrdfs"
+	"go-hep.org/x/hep/xrootd/xrdproto"
+)
+
+type fs struct {
+	pathfs.FileSystem
+	client *client.Client
+	xrdfs  xrdfs.FileSystem
+	root   string
+}
+
+// NewFS returns a new path.FileSystem representing filesystem on the remote XRootD server.
+// client is a client to the remote XRootD server.
+// root is the path to the remote directory to be used as a root directory.
+func NewFS(client *client.Client, root string) pathfs.FileSystem {
+	return &fs{
+		FileSystem: pathfs.NewDefaultFileSystem(),
+		client:     client,
+		xrdfs:      client.FS(),
+		root:       root,
+	}
+}
+
+// GetAttr implements pathfs.FileSystem.GetAttr
+func (fs *fs) GetAttr(name string, ctx *fuse.Context) (*fuse.Attr, fuse.Status) {
+	stat, err := fs.xrdfs.Stat(context.Background(), path.Join(fs.root, name))
+	if serverError, ok := err.(xrdproto.ServerError); ok {
+		if serverError.Code == 3011 {
+			// File does not exists.
+			return nil, fuse.ENOENT
+		}
+	}
+	if err != nil {
+		log.Printf("xrdfuse: error calling stat: %v\n", err)
+		return nil, fuse.EIO
+	}
+
+	var out fuse.Attr
+	out.Size = uint64(stat.Size())
+	out.Mtime = uint64(stat.Mtime)
+	out.Mode = entryStatToMode(stat)
+
+	return &out, fuse.OK
+}
+
+// OpenDir implements pathfs.FileSystem.OpenDir
+func (fs *fs) OpenDir(name string, ctx *fuse.Context) (c []fuse.DirEntry, code fuse.Status) {
+	entries, err := fs.xrdfs.Dirlist(context.Background(), path.Join(fs.root, name))
+	if err != nil {
+		log.Printf("xrdfuse: error calling dirlist: %v\n", err)
+		return nil, fuse.EIO
+	}
+
+	s := make([]fuse.DirEntry, 0, len(entries))
+	for _, entry := range entries {
+		s = append(s, fuse.DirEntry{Name: entry.Name(), Mode: entryStatToMode(entry)})
+	}
+
+	return s, fuse.OK
+}
+
+func convertFlagsToMode(flags uint32) xrdfs.OpenMode {
+	var result xrdfs.OpenMode
+	if flags&uint32(os.O_RDWR) != 0 {
```

Comment:
this would be more readable with a `switch`:
```go
switch {
case flags&uint32(os.O_RDWR) != 0:
case flags&uint32(os.O_RDONLY) != 0:
case flags&uint32(os.O_WRONLY) != 0:
}
```